In [1]:
import numpy as np 
import matplotlib.pyplot as plt 
import pandas as pd 

# Data Tidying and Cleaning Lab
## Reading, tidying and cleaning data. Preparing data for exploration, mining, analysis and learning

In this lab, you'll be working with the Coffee Quality Index dataset, located [here](https://www.kaggle.com/datasets/volpatto/coffee-quality-database-from-cqi). For convenience (and to save trouble in case you can't download files, or someone uploads a newer version), I've provided the dataset in the `data/` folder. The metadata (description) is at the Kaggle link. For this lab, you'll only need `merged_data_cleaned.csv`, as it is the concatenation of the other two datasets.

In this (and the following labs), you'll get several questions and problems. Do your analysis, describe it, use any tools and plots you wish, and answer. You can create any amount of cells you'd like.

Sometimes, the answers will not be unique, and they will depend on how you decide to approach and solve the problem. This is usual - we're doing science after all!

It's a good idea to save your clean dataset after all the work you've done to it.

### Problem 1. Read the dataset (1 point)
This should be self-explanatory. The first column is the index.

1. Reading the dataset using the Pandas library's **read** function and saving it in a variable named **coffe_data**

In [3]:
coffee_data = pd.read_csv("Data/merged_data_cleaned.csv")

2. Calling the variable **coffee_data** in order to show the DataFrame of the set.

In [4]:
coffee_data

Unnamed: 0  Species                     Owner Country.of.Origin  \
0              0  Arabica                 metad plc          Ethiopia   
1              1  Arabica                 metad plc          Ethiopia   
2              2  Arabica  grounds for health admin         Guatemala   
3              3  Arabica       yidnekachew dabessa          Ethiopia   
4              4  Arabica                 metad plc          Ethiopia   
...          ...      ...                       ...               ...   
1334        1334  Robusta               luis robles           Ecuador   
1335        1335  Robusta               luis robles           Ecuador   
1336        1336  Robusta               james moore     United States   
1337        1337  Robusta             cafe politico             India   
1338        1338  Robusta             cafe politico           Vietnam   

                                     Farm.Name Lot.Number            Mill  \
0                                    metad plc        NaN       metad plc   
1                                    metad plc        NaN       metad plc   
2     san marcos barrancas "san cristobal cuch        NaN             NaN   
3        yidnekachew dabessa coffee plantation        NaN         wolensu   
4                                    metad plc        NaN       metad plc   
...                                        ...        ...             ...   
1334                                 robustasa   Lavado 1     our own lab   
1335                                 robustasa   Lavado 3  own laboratory   
1336                           fazenda cazengo        NaN    cafe cazengo   
1337                                       NaN        NaN             NaN   
1338                                       NaN        NaN             NaN   

             ICO.Number                                Company       Altitude  \
0             2014/2015      metad agricultural developmet plc      1950-2200   
1             2014/2015      metad agricultural developmet plc      1950-2200   
2                   NaN                                    NaN  1600 - 1800 m   
3                   NaN  yidnekachew debessa coffee plantation      1800-2200   
4             2014/2015      metad agricultural developmet plc      1950-2200   
...                 ...                                    ...            ...   
1334                NaN                              robustasa            NaN   
1335                NaN                              robustasa             40   
1336                NaN                global opportunity fund     795 meters   
1337  14-1118-2014-0087                          cafe politico            NaN   
1338                NaN                          cafe politico            NaN   

      ...       Color Category.Two.Defects           Expiration  \
0     ...       Green                    0      April 3rd, 2016   
1     ...       Green                    1      April 3rd, 2016   
2     ...         NaN                    0       May 31st, 2011   
3     ...       Green                    2     March 25th, 2016   
4     ...       Green                    2      April 3rd, 2016   
...   ...         ...                  ...                  ...   
1334  ...  Blue-Green                    1   January 18th, 2017   
1335  ...  Blue-Green                    0   January 18th, 2017   
1336  ...         NaN                    6  December 23rd, 2015   
1337  ...       Green                    1    August 25th, 2015   
1338  ...         NaN                    9    August 25th, 2015   

                      Certification.Body  \
0     METAD Agricultural Development plc   
1     METAD Agricultural Development plc   
2           Specialty Coffee Association   
3     METAD Agricultural Development plc   
4     METAD Agricultural Development plc   
...                                  ...   
1334        Specialty Coffee Association   
1335        Specialty Coffee Association   
1336        Specialty Coffee A

### Problem 2. Observations and features (1 point)
How many observations are there? How many features? Which features are numerical, and which are categorical?

**Note:** Think about the _meaning_, not the data types. The dataset hasn't been thoroughly cleaned.

1. In order to get the number of rows (Observations) and columns (features) I use the **shape** function from the Pandas library.

In [6]:
coffee_data.shape

(1339, 44)

2. From the output we can see that there are **1339 observations** and **44 features**

3. Using the **columns** function from the Pandas library in order to explore the 44 features of the set.

In [16]:
coffee_data.columns

Index(['Unnamed: 0', 'Species', 'Owner', 'Country.of.Origin', 'Farm.Name',
       'Lot.Number', 'Mill', 'ICO.Number', 'Company', 'Altitude', 'Region',
       'Producer', 'Number.of.Bags', 'Bag.Weight', 'In.Country.Partner',
       'Harvest.Year', 'Grading.Date', 'Owner.1', 'Variety',
       'Processing.Method', 'Aroma', 'Flavor', 'Aftertaste', 'Acidity', 'Body',
       'Balance', 'Uniformity', 'Clean.Cup', 'Sweetness', 'Cupper.Points',
       'Total.Cup.Points', 'Moisture', 'Category.One.Defects', 'Quakers',
       'Color', 'Category.Two.Defects', 'Expiration', 'Certification.Body',
       'Certification.Address', 'Certification.Contact', 'unit_of_measurement',
       'altitude_low_meters', 'altitude_high_meters', 'altitude_mean_meters'],
      dtype='object')

4. Using the **dtypes** will show us all the columns and their datatypes(The dataset is not clean, so the datatypes might not be correct)

In [17]:
coffee_data.dtypes

Unnamed: 0                 int64
Species                   object
Owner                     object
Country.of.Origin         object
Farm.Name                 object
Lot.Number                object
Mill                      object
ICO.Number                object
Company                   object
Altitude                  object
Region                    object
Producer                  object
Number.of.Bags             int64
Bag.Weight                object
In.Country.Partner        object
Harvest.Year              object
Grading.Date              object
Owner.1                   object
Variety                   object
Processing.Method         object
Aroma                    float64
Flavor                   float64
Aftertaste               float64
Acidity                  float64
Body                     float64
Balance                  float64
Uniformity               float64
Clean.Cup                float64
Sweetness                float64
Cupper.Points            float64
Total.Cup.

### Problem 3. Column manipulation (1 point)
Make the column names more Pythonic (which helps with the quality and... aesthetics). Convert column names to `snake_case`, i.e. `species`, `country_of_origin`, `ico_number`, etc. Try to not do it manually.

1. Defining a function to convert the column names to snake_case (by making every letter lowercase and replacing the spaces or periods(dots) in the names with underscores.

In [19]:
 """
    Convert a column name to snake_case.

    This function takes a column name as input and converts it to snake_case format.
    Snake_case format means all characters are lowercase and spaces or periods are replaced with underscores.

    Parameters:
    column_name (str): The name of the column to be converted.

    Returns:
    str: The converted column name in snake_case format.
 """
def convert_to_snake_case(column_name):
    return column_name.lower().replace('.', '_').replace(' ', '_')

2. Using **.lower** will make every letter lowercase.
3. Using **.replace()** replaces a specified letter,symbol or phrase with another specified one.
4. The function returns the converted column names
5. Renaming the columns of the original dataset coffee_data

In [ ]:
coffee_data.rename(columns=lambda col: convert_to_snake_case(col))

6. Using the **.rename** method creates a new DataFrame with renamed columns, based on the provided parameters. It takes the column names as an input and it applies the conversion function to every single column name in the dataset, and by this way renaming every column name and creating a new dataframe with the new converted column names.
7. Saving the new DataFrame in a new variable, in order to preserve the original one in case it's needed

In [20]:
coffee_data_converted_names = coffee_data.rename(columns=lambda col: convert_to_snake_case(col))

8. Both DataFrames can now be called and used and no information is lost.

In [24]:
coffee_data # ORIGINAL DATAFRAME

Unnamed: 0  Species                     Owner Country.of.Origin  \
0              0  Arabica                 metad plc          Ethiopia   
1              1  Arabica                 metad plc          Ethiopia   
2              2  Arabica  grounds for health admin         Guatemala   
3              3  Arabica       yidnekachew dabessa          Ethiopia   
4              4  Arabica                 metad plc          Ethiopia   
...          ...      ...                       ...               ...   
1334        1334  Robusta               luis robles           Ecuador   
1335        1335  Robusta               luis robles           Ecuador   
1336        1336  Robusta               james moore     United States   
1337        1337  Robusta             cafe politico             India   
1338        1338  Robusta             cafe politico           Vietnam   

                                     Farm.Name Lot.Number            Mill  \
0                                    metad plc        NaN       metad plc   
1                                    metad plc        NaN       metad plc   
2     san marcos barrancas "san cristobal cuch        NaN             NaN   
3        yidnekachew dabessa coffee plantation        NaN         wolensu   
4                                    metad plc        NaN       metad plc   
...                                        ...        ...             ...   
1334                                 robustasa   Lavado 1     our own lab   
1335                                 robustasa   Lavado 3  own laboratory   
1336                           fazenda cazengo        NaN    cafe cazengo   
1337                                       NaN        NaN             NaN   
1338                                       NaN        NaN             NaN   

             ICO.Number                                Company       Altitude  \
0             2014/2015      metad agricultural developmet plc      1950-2200   
1             2014/2015      metad agricultural developmet plc      1950-2200   
2                   NaN                                    NaN  1600 - 1800 m   
3                   NaN  yidnekachew debessa coffee plantation      1800-2200   
4             2014/2015      metad agricultural developmet plc      1950-2200   
...                 ...                                    ...            ...   
1334                NaN                              robustasa            NaN   
1335                NaN                              robustasa             40   
1336                NaN                global opportunity fund     795 meters   
1337  14-1118-2014-0087                          cafe politico            NaN   
1338                NaN                          cafe politico            NaN   

      ...       Color Category.Two.Defects           Expiration  \
0     ...       Green                    0      April 3rd, 2016   
1     ...       Green                    1      April 3rd, 2016   
2     ...         NaN                    0       May 31st, 2011   
3     ...       Green                    2     March 25th, 2016   
4     ...       Green                    2      April 3rd, 2016   
...   ...         ...                  ...                  ...   
1334  ...  Blue-Green                    1   January 18th, 2017   
1335  ...  Blue-Green                    0   January 18th, 2017   
1336  ...         NaN                    6  December 23rd, 2015   
1337  ...       Green                    1    August 25th, 2015   
1338  ...         NaN                    9    August 25th, 2015   

                      Certification.Body  \
0     METAD Agricultural Development plc   
1     METAD Agricultural Development plc   
2           Specialty Coffee Association   
3     METAD Agricultural Development plc   
4     METAD Agricultural Development plc   
...                                  ...   
1334        Specialty Coffee Association   
1335        Specialty Coffee Association   
1336        Specialty Coffee A

In [25]:
coffee_data_converted_names # DATAFRAME WITH CONVERTED COLUMN NAMES

unnamed:_0  species                     owner country_of_origin  \
0              0  Arabica                 metad plc          Ethiopia   
1              1  Arabica                 metad plc          Ethiopia   
2              2  Arabica  grounds for health admin         Guatemala   
3              3  Arabica       yidnekachew dabessa          Ethiopia   
4              4  Arabica                 metad plc          Ethiopia   
...          ...      ...                       ...               ...   
1334        1334  Robusta               luis robles           Ecuador   
1335        1335  Robusta               luis robles           Ecuador   
1336        1336  Robusta               james moore     United States   
1337        1337  Robusta             cafe politico             India   
1338        1338  Robusta             cafe politico           Vietnam   

                                     farm_name lot_number            mill  \
0                                    metad plc        NaN       metad plc   
1                                    metad plc        NaN       metad plc   
2     san marcos barrancas "san cristobal cuch        NaN             NaN   
3        yidnekachew dabessa coffee plantation        NaN         wolensu   
4                                    metad plc        NaN       metad plc   
...                                        ...        ...             ...   
1334                                 robustasa   Lavado 1     our own lab   
1335                                 robustasa   Lavado 3  own laboratory   
1336                           fazenda cazengo        NaN    cafe cazengo   
1337                                       NaN        NaN             NaN   
1338                                       NaN        NaN             NaN   

             ico_number                                company       altitude  \
0             2014/2015      metad agricultural developmet plc      1950-2200   
1             2014/2015      metad agricultural developmet plc      1950-2200   
2                   NaN                                    NaN  1600 - 1800 m   
3                   NaN  yidnekachew debessa coffee plantation      1800-2200   
4             2014/2015      metad agricultural developmet plc      1950-2200   
...                 ...                                    ...            ...   
1334                NaN                              robustasa            NaN   
1335                NaN                              robustasa             40   
1336                NaN                global opportunity fund     795 meters   
1337  14-1118-2014-0087                          cafe politico            NaN   
1338                NaN                          cafe politico            NaN   

      ...       color category_two_defects           expiration  \
0     ...       Green                    0      April 3rd, 2016   
1     ...       Green                    1      April 3rd, 2016   
2     ...         NaN                    0       May 31st, 2011   
3     ...       Green                    2     March 25th, 2016   
4     ...       Green                    2      April 3rd, 2016   
...   ...         ...                  ...                  ...   
1334  ...  Blue-Green                    1   January 18th, 2017   
1335  ...  Blue-Green                    0   January 18th, 2017   
1336  ...         NaN                    6  December 23rd, 2015   
1337  ...       Green                    1    August 25th, 2015   
1338  ...         NaN                    9    August 25th, 2015   

                      certification_body  \
0     METAD Agricultural Development plc   
1     METAD Agricultural Development plc   
2           Specialty Coffee Association   
3     METAD Agricultural Development plc   
4     METAD Agricultural Development plc   
...                                  ...   
1334        Specialty Coffee Association   
1335        Specialty Coffee Association   
1336        Specialty Coffee A

### Problem 4. Bag weight (1 point)
What's up with the bag weights? Make all necessary changes to the column values. Don't forget to document your methods and assumptions.

1. Using **.unique()** to show all the values of the bag_weight column in order to see what is up with the values

In [27]:
coffee_data_converted_names.bag_weight.unique()

array(['60 kg', '1', '30 kg', '69 kg', '1 kg', '2 kg,lbs', '6', '3 lbs',
       '50 kg', '2 lbs', '100 lbs', '15 kg', '2 kg', '2', '70 kg',
       '19200 kg', '5 lbs', '1 kg,lbs', '6 kg', '0 lbs', '46 kg', '40 kg',
       '20 kg', '34 kg', '1 lbs', '660 kg', '18975 kg', '12000 kg',
       '35 kg', '66 kg', '80 kg', '132 lbs', '5 kg', '25 kg', '59 kg',
       '18000 kg', '150 lbs', '9000 kg', '18 kg', '10 kg', '29 kg',
       '1218 kg', '4 lbs', '0 kg', '13800 kg', '1500 kg', '24 kg',
       '80 lbs', '8 kg', '3 kg', '350 kg', '67 kg', '4 kg', '55 lbs',
       '100 kg', '130 lbs'], dtype=object)

2. The bag_weight column contains various weights in different formats, such as:
- Values in different units ('kg' or 'lbs')
- Mixed unit format (for example "1 kg,lbs")
- Values with missing units(for example "1")
- Values which look way bigger than the rest of the values
- Values that are 0 in kg or in lbs("0 kg" or "0 lbs")

3. Using value_counts() to see the number of occurences of different values in order to decide what to do with the units(for example to standardize the units by converting them to one unit) and to see the occurance of anomaly values to decide whether to keep them and convert them 

In [28]:
coffee_data_converted_names.bag_weight.value_counts()

bag_weight
1 kg        331
60 kg       256
69 kg       200
70 kg       156
2 kg        122
100 lbs      59
30 kg        29
5 lbs        23
6            19
20 kg        14
50 kg        14
10 kg        11
59 kg        10
1 lbs         8
1             7
3 lbs         7
5 kg          7
2 lbs         5
4 lbs         4
80 kg         4
18975 kg      4
0 lbs         3
46 kg         3
29 kg         2
9000 kg       2
25 kg         2
66 kg         2
35 kg         2
12000 kg      2
40 kg         2
6 kg          2
19200 kg      2
15 kg         2
13800 kg      1
100 kg        1
55 lbs        1
4 kg          1
67 kg         1
350 kg        1
3 kg          1
8 kg          1
80 lbs        1
24 kg         1
1500 kg       1
2 kg,lbs      1
0 kg          1
660 kg        1
1218 kg       1
2             1
18 kg         1
150 lbs       1
18000 kg      1
1 kg,lbs      1
132 lbs       1
34 kg         1
130 lbs       1
Name: count, dtype: int64

4. Let's check which unit occurs more (kg's or lbs)

### Problem 5. Dates (1 point)
This should remind you of problem 4 but it's slightly nastier. Fix the harvest years, document the process.

While you're here, fix the expiration dates, and grading dates. Unlike the other column, these should be dates (`pd.to_datetime()` is your friend).

### Problem 6. Countries (1 point)
How many coffees are there with unknown countries of origin? What can you do about them?

### Problem 7. Owners (1 point)
There are two suspicious columns, named `Owner`, and `Owner.1` (they're likely called something different after you solved problem 3). Do something about them. Is there any link to `Producer`?

### Problem 8. Coffee color by country and continent (1 point)
Create a table which shows how many coffees of each color are there in every country. Leave the missing values as they are.

**Note:** If you ask me, countries should be in rows, I prefer long tables much better than wide ones.

Now do the same for continents. You know what continent each country is located in.

### Problem 9. Ratings (1 point)
The columns `Aroma`, `Flavor`, etc., up to `Moisture` represent subjective ratings. Explore them. Show the means and range; draw histograms and / or boxplots as needed. You can even try correlations if you want. What's up with all those ratings?

### Problem 10. High-level errors (1 point)
Check the countries against region names, altitudes, and companies. Are there any discrepancies (e.g. human errors, like a region not matching the country)? Take a look at the (cleaned) altitudes; there has been a lot of preprocessing done to them. Was it done correctly?

### * Problem 11. Clean and explore at will
The dataset claimed to be clean, but we were able to discover a lot of things to fix and do better.

Play around with the data as much as you wish, and if you find variables to tidy up and clean - by all means, do that!